contracts/Market.sol

```diff
@@ -0,0 +1,857 @@
+   * @return ids An array of the removal token ids being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   * @return suppliers The address of the supplier who owns each corresponding removal token.
+   */
+  function _allocateSupplyRoundRobin(uint256 certificateAmount)
+    private
+    returns (
+      uint256,
+      uint256[] memory,
+      uint256[] memory,
+      address[] memory
+    )
+  {
+    uint256 remainingAmountToFill = certificateAmount;
+    uint256 numberOfActiveRemovalsInMarket = _removal
+      .numberOfTokensOwnedByAddress(
+        address(this) // todo are we making calls like this throughout the contract flow? if so can we pass values down?
+      );
+    uint256[] memory ids = new uint256[](numberOfActiveRemovalsInMarket);
+    uint256[] memory amounts = new uint256[](numberOfActiveRemovalsInMarket);
+    address[] memory suppliers = new address[](numberOfActiveRemovalsInMarket);
+    uint256 numberOfRemovalsForOrder = 0;
+    // TODO (Gas Optimization): Declare variables outside of loop
+    for (
+      uint256 i = 0;
+      i < numberOfActiveRemovalsInMarket;
+      i++ // todo ++i consistency
+    ) {
+      uint256 removalId = _activeSupply[_currentSupplierAddress]
+        .getNextRemovalForSale();
+      uint256 removalAmount = _removal.balanceOf(address(this), removalId); // todo batch
+      if (remainingAmountToFill < removalAmount) {
+        /**
+         * The order is complete, not fully using up this removal, don't increment currentSupplierAddress,
+         * don't check about removing active supplier.
+         */
+        ids[numberOfRemovalsForOrder] = removalId;
+        amounts[numberOfRemovalsForOrder] = remainingAmountToFill;
+        suppliers[numberOfRemovalsForOrder] = _currentSupplierAddress;
+        remainingAmountToFill = 0;
+      } else {
+        /**
+         * We will use up this removal while completing the order, move on to next one.
+         */
+        ids[numberOfRemovalsForOrder] = removalId;
+        amounts[numberOfRemovalsForOrder] = removalAmount; // this removal is getting used up
+        suppliers[numberOfRemovalsForOrder] = _currentSupplierAddress;
+        remainingAmountToFill -= removalAmount;
+        _activeSupply[_currentSupplierAddress].removeRemoval(removalId);
+        if (_activeSupply[_currentSupplierAddress].isRemovalQueueEmpty()) {
+          /**
+           * If the supplier is out of supply, remove them from the active suppliers.
+           */
+          _removeActiveSupplier(_currentSupplierAddress);
+        } else if (
+          /**
+           *  If the supplier is the only supplier remaining with supply, don't bother incrementing.
+           */
+          _suppliersInRoundRobinOrder[_currentSupplierAddress]
+            .nextSupplierAddress != _currentSupplierAddress
+        ) {
+          _incrementCurrentSupplierAddress();
+        }
+      }
```

Comment:
I'd like to see the following code refactored so that the maintenance of the queue and round robin pointers lives in separate functions that are well documented.

Reply:
tracked